invesalius/gui/widgets/canvas_renderer.py

```diff
@@ -67,59 +100,59 @@ def __init__(self, viewer, evt_renderer, canvas_renderer, orientation=None):
         self.canvas_renderer = canvas_renderer
         self.evt_renderer = evt_renderer
         self._size = self.canvas_renderer.GetSize()
-        self.draw_list = []
-        self._ordered_draw_list = []
+        self.draw_list: List[CanvasHandlerBase] = []
+        self._ordered_draw_list: List[Tuple[int, CanvasHandlerBase]] = []
         self.orientation = orientation
-        self.gc = None
-        self.last_cam_modif_time = -1
-        self.modified = True
-        self._drawn = False
+        self.gc: Optional[wx.GraphicsContext] = None
+        self.last_cam_modif_time: int = -1
+        self.modified: bool = True
+        self._drawn: bool = False
         self._init_canvas()
 
         self._over_obj = None
         self._drag_obj = None
         self._selected_obj = None
 
-        self._callback_events = {
-            'LeftButtonPressEvent': [],
-            'LeftButtonReleaseEvent': [],
-            'LeftButtonDoubleClickEvent': [],
-            'MouseMoveEvent': [],
+        self._callback_events: Dict[str, List[Callable]] = {
+            "LeftButtonPressEvent": [],
+            "LeftButtonReleaseEvent": [],
+            "LeftButtonDoubleClickEvent": [],
+            "MouseMoveEvent": [],
         }
 
         self._bind_events()
 
-    def _bind_events(self):
+    def _bind_events(self) -> None:
         iren = self.viewer.interactor
         iren.Bind(wx.EVT_MOTION, self.OnMouseMove)
         iren.Bind(wx.EVT_LEFT_DOWN, self.OnLeftButtonPress)
         iren.Bind(wx.EVT_LEFT_UP, self.OnLeftButtonRelease)
         iren.Bind(wx.EVT_LEFT_DCLICK, self.OnDoubleClick)
         self.canvas_renderer.AddObserver("StartEvent", self.OnPaint)
 
-    def subscribe_event(self, event, callback):
+    def subscribe_event(self, event, callback) -> None:
         ref = WeakMethod(callback)
         self._callback_events[event].append(ref)
 
-    def unsubscribe_event(self, event, callback):
+    def unsubscribe_event(self, event, callback) -> None:
         for n, cb in enumerate(self._callback_events[event]):
             if cb() == callback:
-                print('removed')
+                print("removed")
                 self._callback_events[event].pop(n)
                 return
 
-    def propagate_event(self, root, event):
-        print('propagating', event.event_name, 'from', root)
+    def propagate_event(self, root, event) -> None:
```

Comment:
root: CanvasObjects
event: CanvasEvent

Reply:
Ok but that would mean that I have to change CanvasObjects to be this definition:
```python
class CanvasObjects(Protocol):
    parent: Self | None
    def is_over(self, x: int, y: int) -> Self | None: ...
    def draw_to_canvas(self, gc: wx.GraphicsContext, canvas: CanvasRendererCTX) -> None: ...
```

Reply:
I didn't understand this.

Reply:
I am sorry if I wasn't clear. I meant I had to add a `parent` attribute to the `CanvasObjects` Protocol since it was used inside the function.

Reply:
Ok